<launch>
  <arg name="map_size_x_"/>
  <arg name="map_size_y_"/>
  <arg name="map_size_z_"/>

  <arg name="odometry_topic"/>
  <arg name="camera_pose_topic"/>
  <arg name="depth_topic"/>
  <arg name="cloud_topic"/>

  <arg name="cx"/>
  <arg name="cy"/>
  <arg name="fx"/>
  <arg name="fy"/>

  <arg name="max_vel"/>
  <arg name="max_acc"/>
  <arg name="planning_horizon"/>

  <arg name="max_vel_yaw"/>

  <arg name="ld_smooth"/>
  <arg name="ld_dist"/>
  <arg name="ld_feasi"/>
  <arg name="ld_start"/>
  <arg name="ld_end"/>
  <arg name="ld_waypt"/>
  <arg name="ld_time"/>
  <arg name="ld_yaw_feasi"/>
  <arg name="ld_parallax"/>
  <arg name="ld_vertical_visibility"/>
  <arg name="ld_yaw_covisibility"/> 


  <arg name="init_x"/>
  <arg name="init_y"/>
  <arg name="init_z"/>
  <arg name="init_yaw"/>


  <arg name="mode"/>
  <arg name="world_frame" value="world"/>
  <arg name="sensor_frame" value="camera"/>

  <arg name="map_min_x"/>
  <arg name="map_min_y"/>
  <arg name="map_min_z"/>
  <arg name="map_max_x"/>
  <arg name="map_max_y"/>
  <arg name="map_max_z"/>

  <arg name="box_min_x"/>
  <arg name="box_min_y"/>
  <arg name="box_min_z"/>
  <arg name="box_max_x"/>
  <arg name="box_max_y"/>
  <arg name="box_max_z"/>

  <arg name="vbox_min_x"/>
  <arg name="vbox_min_y"/>
  <arg name="vbox_min_z"/>
  <arg name="vbox_max_x"/>
  <arg name="vbox_max_y"/>
  <arg name="vbox_max_z"/>


  <arg name="flight_type"/>

  <!-- main node -->
  <node pkg="ego_planner" name="ego_planner_node" type="ego_planner_node" output="screen">
    <remap from="/odom_world" to="$(arg odometry_topic)"/>
    <remap from="/grid_map/odom" to="$(arg odometry_topic)"/>
    <remap from="/grid_map/cloud" to="$(arg cloud_topic)"/>
    <remap from = "/grid_map/pose"   to = "$(arg camera_pose_topic)"/>
    <remap from = "/grid_map/depth" to = "$(arg depth_topic)"/>

    <!-- planning fsm -->
    <param name="fsm/flight_type" value="$(arg flight_type)" type="int"/>
    <param name="fsm/thresh_replan" value="1.5" type="double"/>
    <param name="fsm/thresh_no_replan" value="2.0" type="double"/>
    <param name="fsm/planning_horizon" value="$(arg planning_horizon)" type="double"/> <!--always set to 1.5 times grater than sensing horizen-->
    <param name="fsm/planning_horizen_time" value="3" type="double"/>
    <param name="fsm/emergency_time_" value="1.0" type="double"/>

    <param name="grid_map/resolution"      value="0.1" />
    <param name="grid_map/map_size_x"   value="$(arg map_size_x_)" />
    <param name="grid_map/map_size_y"   value="$(arg map_size_y_)" />
    <param name="grid_map/map_size_z"   value="$(arg map_size_z_)" />
    <param name="grid_map/local_update_range_x"  value="5.5" />
    <param name="grid_map/local_update_range_y"  value="5.5" />
    <param name="grid_map/local_update_range_z"  value="4.5" />
    <param name="grid_map/obstacles_inflation"     value="0.099" />
    <param name="grid_map/local_map_margin" value="30"/>
    <param name="grid_map/ground_height"        value="-0.01"/>
    <!-- camera parameter -->
    <param name="grid_map/cx" value="$(arg cx)"/>
    <param name="grid_map/cy" value="$(arg cy)"/>
    <param name="grid_map/fx" value="$(arg fx)"/>
    <param name="grid_map/fy" value="$(arg fy)"/>
    <!-- depth filter -->
    <param name="grid_map/use_depth_filter" value="true"/>
    <param name="grid_map/depth_filter_tolerance" value="0.15"/>
    <param name="grid_map/depth_filter_maxdist"   value="5.0"/>
    <param name="grid_map/depth_filter_mindist"   value="0.2"/>
    <param name="grid_map/depth_filter_margin"    value="1"/>
    <param name="grid_map/k_depth_scaling_factor" value="1000.0"/>
    <param name="grid_map/skip_pixel" value="2"/>
    <!-- local fusion -->
    <param name="grid_map/p_hit"  value="0.65"/>
    <param name="grid_map/p_miss" value="0.35"/>
    <param name="grid_map/p_min"  value="0.12"/>
    <param name="grid_map/p_max"  value="0.90"/>
    <param name="grid_map/p_occ"  value="0.80"/>
    <param name="grid_map/min_ray_length" value="0.1"/>
    <param name="grid_map/max_ray_length" value="4.5"/>

    <param name="grid_map/virtual_ceil_height"   value="2.5"/>
    <param name="grid_map/visualization_truncate_height"   value="2.4"/>
    <param name="grid_map/show_occ_time"  value="false"/>
    <param name="grid_map/pose_type"     value="2"/>
    <param name="grid_map/frame_id"      value="world"/>

  <!-- planner manager -->
    <param name="manager/max_vel" value="$(arg max_vel)" type="double"/>
    <param name="manager/max_acc" value="$(arg max_acc)" type="double"/>
    <param name="manager/max_jerk" value="4" type="double"/>
    <param name="manager/control_points_distance" value="0.4" type="double"/>
    <param name="manager/feasibility_tolerance" value="0.05" type="double"/>
    <param name="manager/planning_horizon" value="$(arg planning_horizon)" type="double"/>

<!--ego 优化-->
    <param name="optimization/lambda_smooth" value="1.0" type="double"/>
    <param name="optimization/lambda_collision" value="0.5" type="double"/>
    <param name="optimization/lambda_feasibility" value="0.1" type="double"/>
    <param name="optimization/lambda_fitness" value="1.0" type="double"/>
    <param name="optimization/dist0" value="0.5" type="double"/>
    <param name="optimization/swarm_clearance" value="0.5" type="double"/>
    <param name="optimization/max_vel" value="$(arg max_vel)" type="double"/>
    <param name="optimization/max_acc" value="$(arg max_acc)" type="double"/>

    <param name="bspline/limit_vel" value="$(arg max_vel)" type="double"/>
    <param name="bspline/limit_acc" value="$(arg max_acc)" type="double"/>
    <param name="bspline/limit_ratio" value="1.1" type="double"/>


    <!-- Yaw planner utils -->
    <param name="yaw_initial/yaw_diff" value="$(eval 30 * 3.1415926 / 180.0)" type="double"/>
    <param name="yaw_initial/half_vert_num" value="5" type="int"/>
    <param name="yaw_initial/max_yaw_rate" value="$(eval 90 * 3.1415926 / 180.0)" type="double"/>
    <param name="yaw_initial/w" value="0.0" type="double"/>
    <param name="yaw_initial/w_forwarding" value="0.5" type="double"/>
    <param name="yaw_initial/weight_type" value="1" type="double"/>

    <!-- <param name="optimization/ld_smooth" value="20.0" type="double"/> -->
    <param name="optimization/ld_smooth" value="$(arg ld_smooth)" type="double"/>
    <param name="optimization/ld_dist" value="$(arg ld_dist)" type="double"/>
    <param name="optimization/ld_feasi" value="$(arg ld_feasi)" type="double"/>
    <param name="optimization/ld_start" value="$(arg ld_start)" type="double"/>
    <param name="optimization/ld_end" value="$(arg ld_end)" type="double"/>
    <!-- <param name="optimization/ld_guide" value="1.5" type="double"/> -->
    <param name="optimization/ld_waypt" value="$(arg ld_waypt)" type="double"/>
    <!-- <param name="optimization/ld_view" value="0.0" type="double"/> -->
    <param name="optimization/ld_time" value="$(arg ld_time)" type="double"/>
    <!-- <param name="optimization/ld_info" value="1.0" type="double"/> -->
    <param name="optimization/ld_yaw_feasi" value="$(arg ld_yaw_feasi)" type="double"/>
    <param name="optimization/ld_parallax" value="$(arg ld_parallax)" type="double"/>
    <param name="optimization/ld_vertical_visibility" value="$(arg ld_vertical_visibility)" type="double"/>
    <param name="optimization/ld_yaw_covisibility" value="$(arg ld_yaw_covisibility)" type="double"/>

    <param name="optimization/max_vel_yaw" value="$(arg max_vel_yaw)" type="double"/>

    <param name="optimization/algorithm1" value="15" type="int"/>
    <param name="optimization/algorithm2" value="11" type="int"/>
    <param name="optimization/max_iteration_num1" value="2" type="int"/>
    <param name="optimization/max_iteration_num2" value="2000" type="int"/>
    <param name="optimization/max_iteration_num3" value="200" type="int"/>
    <param name="optimization/max_iteration_num4" value="200000" type="int"/>

    <param name="optimization/max_iteration_time1" value="0.0001" type="double"/>
    <param name="optimization/max_iteration_time2" value="0.01" type="double"/>
    <param name="optimization/max_iteration_time3" value="0.003" type="double"/>
    <param name="optimization/max_iteration_time4" value="2.0" type="double"/>

    <param name="optimization/parallax/estimator_freq" value="10.0" type="double"/>
    <!-- <param name="optimization/parallax/max_parallax" value="$(eval 7 * 3.1415926 / 180.0)" type="double"/>
    <param name="optimization/parallax/pot_a" value="15.0" type="double"/> -->
    <param name="optimization/parallax/max_parallax" value="$(eval 10 * 3.1415926 / 180.0)" type="double"/>
    <param name="optimization/parallax/pot_a" value="20.0" type="double"/>


    <param name="voxel_mapping/p_hit" value="0.65" type="double"/>
    <param name="voxel_mapping/p_miss" value="0.35" type="double"/>
    <param name="voxel_mapping/p_min" value="0.12" type="double"/>
    <param name="voxel_mapping/p_max" value="0.90" type="double"/>
    <param name="voxel_mapping/p_occ" value="0.80" type="double"/>
    <param name="voxel_mapping/virtual_ceil_height" value="-10" type="int"/>
    <param name="voxel_mapping/map_min_x" value="$(arg map_min_x)" type="double"/>
    <param name="voxel_mapping/map_min_y" value="$(arg map_min_y)" type="double"/>
    <param name="voxel_mapping/map_min_z" value="$(arg map_min_z)" type="double"/>
    <param name="voxel_mapping/map_max_x" value="$(arg map_max_x)" type="double"/>
    <param name="voxel_mapping/map_max_y" value="$(arg map_max_y)" type="double"/>
    <param name="voxel_mapping/map_max_z" value="$(arg map_max_z)" type="double"/>
    <param name="voxel_mapping/box_min_x" value="$(arg box_min_x)" type="double"/>
    <param name="voxel_mapping/box_min_y" value="$(arg box_min_y)" type="double"/>
    <param name="voxel_mapping/box_min_z" value="$(arg box_min_z)" type="double"/>
    <param name="voxel_mapping/box_max_x" value="$(arg box_max_x)" type="double"/>
    <param name="voxel_mapping/box_max_y" value="$(arg box_max_y)" type="double"/>
    <param name="voxel_mapping/box_max_z" value="$(arg box_max_z)" type="double"/>
    <param name="voxel_mapping/vbox_min_x" value="$(arg vbox_min_x)" type="double"/>
    <param name="voxel_mapping/vbox_min_y" value="$(arg vbox_min_y)" type="double"/>
    <param name="voxel_mapping/vbox_min_z" value="$(arg vbox_min_z)" type="double"/>
    <param name="voxel_mapping/vbox_max_x" value="$(arg vbox_max_x)" type="double"/>
    <param name="voxel_mapping/vbox_max_y" value="$(arg vbox_max_y)" type="double"/>
    <param name="voxel_mapping/vbox_max_z" value="$(arg vbox_max_z)" type="double"/>




    <param name="voxel_mapping/mode" value="$(arg mode)" type="string"/>

    <param name="transformer/world_frame" value="$(arg world_frame)" type="string"/>
    <param name="transformer/sensor_frame" value="$(arg sensor_frame)" type="string"/>
    <param name="transformer/timestamp_tolerance" value="1e-3" type="double"/>
    <param name="transformer/verbose" value="false" type="bool"/>
    <param name="transformer/pose_topic_type" value="pose" type="string"/>


  </node>

</launch>